File: ros2-bridge/rosserial2/control.py
"""Control-plane opcode codec.

Control frames are frames whose ``msg_id`` is 0. The payload begins
with a one-byte opcode followed by an opcode-specific body. See
``docs/wire-protocol.md`` for the table.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

DEVICE_ID_LEN = 16
FW_HASH_LEN = 20


class Opcode(IntEnum):
    HELLO = 0x01
    HELLO_ACK = 0x02
    ADVERTISE = 0x03
    ADVERTISE_ACK = 0x04
    PING = 0x05
    PONG = 0x06
    LOG = 0x07
    BYE = 0xFF


class Direction(IntEnum):
    PUBLISH = 0
    SUBSCRIBE = 1


class ControlError(ValueError):
    """Raised when a control payload cannot be parsed or built."""


# --- HELLO -----------------------------------------------------------


@dataclass(frozen=True)
class Hello:
    proto_ver: int
    device_id: bytes  # exactly DEVICE_ID_LEN bytes
    fw_hash: bytes  # exactly FW_HASH_LEN bytes
    max_payload: int

    def __post_init__(self) -> None:
        if not 0 <= self.proto_ver <= 0xFF:
            raise ControlError(f"proto_ver out of range: {self.proto_ver}")
        if len(self.device_id) != DEVICE_ID_LEN:
            raise ControlError(f"device_id must be {DEVICE_ID_LEN} bytes")
        if len(self.fw_hash) != FW_HASH_LEN:
            raise ControlError(f"fw_hash must be {FW_HASH_LEN} bytes")
        if not 0 <= self.max_payload <= 0xFFFF:
            raise ControlError(f"max_payload out of range: {self.max_payload}")

    def encode(self) -> bytes:
        return (
            bytes([Opcode.HELLO, self.proto_ver])
            + self.device_id
            + self.fw_hash
            + self.max_payload.to_bytes(2, "little")
        )


def _decode_hello(body: bytes) -> Hello:
    expected = 1 + DEVICE_ID_LEN + FW_HASH_LEN + 2
    if len(body) != expected:
        raise ControlError(f"HELLO body length: got {len(body)}, expected {expected}")
    proto_ver = body[0]
    off = 1
    device_id = body[off : off + DEVICE_ID_LEN]
    off += DEVICE_ID_LEN
    fw_hash = body[off : off + FW_HASH_LEN]
    off += FW_HASH_LEN
    max_payload = int.from_bytes(body[off : off + 2], "little")
    return Hello(
        proto_ver=proto_ver,
        device_id=bytes(device_id),
        fw_hash=bytes(fw_hash),
        max_payload=max_payload,
    )


# --- HELLO_ACK -------------------------------------------------------


@dataclass(frozen=True)
class HelloAck:
    proto_ver: int
    accepted: int
    session_id: int

    def __post_init__(self) -> None:
        if not 0 <= self.proto_ver <= 0xFF:
            raise ControlError(f"proto_ver out of range: {self.proto_ver}")
        if not 0 <= self.accepted <= 0xFF:
            raise ControlError(f"accepted out of range: {self.accepted}")
        if not 0 <= self.session_id <= 0xFFFFFFFF:
            raise ControlError(f"session_id out of range: {self.session_id}")

    def encode(self) -> bytes:
        return (
            bytes([Opcode.HELLO_ACK, self.proto_ver, self.accepted])
            + self.session_id.to_bytes(4, "little")
        )


def _decode_hello_ack(body: bytes) -> HelloAck:
    if len(body) != 6:
        raise ControlError(f"HELLO_ACK body length: got {len(body)}, expected 6")
    return HelloAck(
        proto_ver=body[0],
        accepted=body[1],
        session_id=int.from_bytes(body[2:6], "little"),
    )


# --- ADVERTISE -------------------------------------------------------


@dataclass(frozen=True)
class Advertise:
    topic_id: int
    direction: Direction
    type_str: str  # ROS2 canonical, e.g. "std_msgs/msg/String"
    name: str

    def __post_init__(self) -> None:
        if not 1 <= self.topic_id <= 0xFF:
            raise ControlError(f"topic_id must be 1..255, got {self.topic_id}")
        if not 0 <= int(self.direction) <= 1:
            raise ControlError(f"direction invalid: {self.direction}")
        type_bytes = self.type_str.encode("utf-8")
        name_bytes = self.name.encode("utf-8")
        if len(type_bytes) > 0xFF:
            raise ControlError("type_str too long (>255 bytes)")
        if len(name_bytes) > 0xFF:
            raise ControlError("name too long (>255 bytes)")
        if not type_bytes:
            raise ControlError("type_str must not be empty")
        if not name_bytes:
            raise ControlError("name must not be empty")

    def encode(self) -> bytes:
        type_bytes = self.type_str.encode("utf-8")
        name_bytes = self.name.encode("utf-8")
        return (
            bytes(
                [
                    Opcode.ADVERTISE,
                    self.topic_id,
                    int(self.direction),
                    len(type_bytes),
                ]
            )
            + type_bytes
            + bytes([len(name_bytes)])
            + name_bytes
        )


def _decode_advertise(body: bytes) -> Advertise:
    if len(body) < 4:
        raise ControlError(f"ADVERTISE body too short: {len(body)}")
    topic_id = body[0]
    direction = body[1]
    if direction not in (0, 1):
        raise ControlError(f"ADVERTISE bad direction: {direction}")
    type_len = body[2]
    off = 3
    if len(body) < off + type_len + 1:
        raise ControlError("ADVERTISE truncated in type_str")
    type_str = body[off : off + type_len].decode("utf-8")
    off += type_len
    name_len = body[off]
    off += 1
    if len(body) != off + name_len:
        raise ControlError(
            f"ADVERTISE size mismatch: trailing {len(body) - (off + name_len)} bytes"
        )
    name = body[off : off + name_len].decode("utf-8")
    return Advertise(
        topic_id=topic_id,
        direction=Direction(direction),
        type_str=type_str,
        name=name,
    )


# --- ADVERTISE_ACK ---------------------------------------------------


@dataclass(frozen=True)
class AdvertiseAck:
    topic_id: int
    accepted: int

    def __post_init__(self) -> None:
        if not 1 <= self.topic_id <= 0xFF:
            raise ControlError(f"topic_id must be 1..255, got {self.topic_id}")
        if not 0 <= self.accepted <= 0xFF:
            raise ControlError(f"accepted out of range: {self.accepted}")

    def encode(self) -> bytes:
        return bytes([Opcode.ADVERTISE_ACK, self.topic_id, self.accepted])


def _decode_advertise_ack(body: bytes) -> AdvertiseAck:
    if len(body) != 2:
        raise ControlError(f"ADVERTISE_ACK body length: got {len(body)}, expected 2")
    return AdvertiseAck(topic_id=body[0], accepted=body[1])


# --- PING / PONG -----------------------------------------------------


@dataclass(frozen=True)
class Ping:
    nonce: int

    def __post_init__(self) -> None:
        if not 0 <= self.nonce <= 0xFFFFFFFF:
            raise ControlError(f"nonce out of range: {self.nonce}")

    def encode(self) -> bytes:
        return bytes([Opcode.PING]) + self.nonce.to_bytes(4, "little")


@dataclass(frozen=True)
class Pong:
    nonce: int

    def __post_init__(self) -> None:
        if not 0 <= self.nonce <= 0xFFFFFFFF:
            raise ControlError(f"nonce out of range: {self.nonce}")

    def encode(self) -> bytes:
        return bytes([Opcode.PONG]) + self.nonce.to_bytes(4, "little")


def _decode_ping(body: bytes) -> Ping:
    if len(body) != 4:
        raise ControlError(f"PING body length: got {len(body)}, expected 4")
    return Ping(nonce=int.from_bytes(body, "little"))


def _decode_pong(body: bytes) -> Pong:
    if len(body) != 4:
        raise ControlError(f"PONG body length: got {len(body)}, expected 4")
    return Pong(nonce=int.from_bytes(body, "little"))


# --- LOG -------------------------------------------------------------


@dataclass(frozen=True)
class Log:
    level: int
    message: str

    def __post_init__(self) -> None:
        if not 0 <= self.level <= 0xFF:
            raise ControlError(f"level out of range: {self.level}")

    def encode(self) -> bytes:
        return bytes([Opcode.LOG, self.level]) + self.message.encode("utf-8")


def _decode_log(body: bytes) -> Log:
    if len(body) < 1:
        raise ControlError("LOG body too short")
    return Log(level=body[0], message=body[1:].decode("utf-8", errors="replace"))


# --- BYE -------------------------------------------------------------


@dataclass(frozen=True)
class Bye:
    reason: int

    def __post_init__(self) -> None:
        if not 0 <= self.reason <= 0xFF:
            raise ControlError(f"reason out of range: {self.reason}")

    def encode(self) -> bytes:
        return bytes([Opcode.BYE, self.reason])


def _decode_bye(body: bytes) -> Bye:
    if len(body) != 1:
        raise ControlError(f"BYE body length: got {len(body)}, expected 1")
    return Bye(reason=body[0])


# --- dispatch --------------------------------------------------------

ControlMessage = Hello | HelloAck | Advertise | AdvertiseAck | Ping | Pong | Log | Bye

_DECODERS = {
    Opcode.HELLO: _decode_hello,
    Opcode.HELLO_ACK: _decode_hello_ack,
    Opcode.ADVERTISE: _decode_advertise,
    Opcode.ADVERTISE_ACK: _decode_advertise_ack,
    Opcode.PING: _decode_ping,
    Opcode.PONG: _decode_pong,
    Opcode.LOG: _decode_log,
    Opcode.BYE: _decode_bye,
}


def decode_control(payload: bytes) -> ControlMessage:
    """Decode a control-frame payload (opcode byte + body)."""
    if not payload:
        raise ControlError("empty control payload")
    op = payload[0]
    try:
        opcode = Opcode(op)
    except ValueError as exc:
        raise ControlError(f"unknown opcode: {op:#04x}") from exc
    decoder = _DECODERS[opcode]
    return decoder(payload[1:])


def encode_control(msg: ControlMessage) -> bytes:
    """Encode a control message to a frame payload."""
    return msg.encode()

File: web-flasher/js/control.js
// Control-plane opcodes. Mirrors ros2-bridge/rosserial2/control.py
// for the opcodes the browser console needs to recognize.
//
// The console is decode-leaning: it parses HELLO/HELLO_ACK,
// ADVERTISE/ADVERTISE_ACK, PING/PONG, LOG, BYE. The encode side
// covers the bare minimum the console needs (HELLO_ACK, PONG) so it
// can play the bridge role for the local handshake demo.

export const Opcode = Object.freeze({
    HELLO: 0x01,
    HELLO_ACK: 0x02,
    ADVERTISE: 0x03,
    ADVERTISE_ACK: 0x04,
    PING: 0x05,
    PONG: 0x06,
    LOG: 0x07,
    BYE: 0xff,
});

export const Direction = Object.freeze({ PUBLISH: 0, SUBSCRIBE: 1 });

export const DEVICE_ID_LEN = 16;
export const FW_HASH_LEN = 20;

export class ControlError extends Error {}

function le16(buf, off) { return buf[off] | (buf[off + 1] << 8); }
function le32(buf, off) {
    return ((buf[off] |
        (buf[off + 1] << 8) |
        (buf[off + 2] << 16) |
        (buf[off + 3] << 24)) >>> 0);
}
function bytesToHex(buf) {
    return Array.from(buf, b => b.toString(16).padStart(2, "0")).join("");
}

/** Decode a control payload. Returns {opcode, ...fields} or throws. */
export function decodeControl(payload) {
    if (!(payload instanceof Uint8Array) || payload.length === 0) {
        throw new ControlError("empty control payload");
    }
    const op = payload[0];
    const body = payload.subarray(1);
    switch (op) {
        case Opcode.HELLO: {
            const expected = 1 + DEVICE_ID_LEN + FW_HASH_LEN + 2;
            if (body.length !== expected) {
                throw new ControlError(`HELLO body length: ${body.length}`);
            }
            return {
                opcode: Opcode.HELLO,
                protoVer: body[0],
                deviceId: body.subarray(1, 1 + DEVICE_ID_LEN),
                fwHash: body.subarray(1 + DEVICE_ID_LEN, 1 + DEVICE_ID_LEN + FW_HASH_LEN),
                maxPayload: le16(body, 1 + DEVICE_ID_LEN + FW_HASH_LEN),
            };
        }
        case Opcode.HELLO_ACK: {
            if (body.length !== 6) throw new ControlError("HELLO_ACK body length");
            return {
                opcode: Opcode.HELLO_ACK,
                protoVer: body[0],
                accepted: body[1],
                sessionId: le32(body, 2),
            };
        }
        case Opcode.ADVERTISE: {
            if (body.length < 3) throw new ControlError("ADVERTISE body too short");
            const topicId = body[0];
            const direction = body[1];
            if (direction > 1) throw new ControlError("ADVERTISE bad direction");
            const typeLen = body[2];
            if (body.length < 3 + typeLen + 1) {
                throw new ControlError("ADVERTISE truncated in type");
            }
            const typeStr = new TextDecoder().decode(body.subarray(3, 3 + typeLen));
            const nameLen = body[3 + typeLen];
            if (body.length !== 4 + typeLen + nameLen) {
                throw new ControlError("ADVERTISE size mismatch");
            }
            const name = new TextDecoder().decode(body.subarray(4 + typeLen));
            return { opcode: Opcode.ADVERTISE, topicId, direction, typeStr, name };
        }
        case Opcode.ADVERTISE_ACK: {
            if (body.length !== 2) throw new ControlError("ADVERTISE_ACK body length");
            return { opcode: Opcode.ADVERTISE_ACK, topicId: body[0], accepted: body[1] };
        }
        case Opcode.PING:
        case Opcode.PONG: {
            if (body.length !== 4) throw new ControlError("PING/PONG body length");
            return { opcode: op, nonce: le32(body, 0) };
        }
        case Opcode.LOG: {
            if (body.length < 1) throw new ControlError("LOG body too short");
            return {
                opcode: Opcode.LOG,
                level: body[0],
                message: new TextDecoder().decode(body.subarray(1)),
            };
        }
        case Opcode.BYE: {
            if (body.length !== 1) throw new ControlError("BYE body length");
            return { opcode: Opcode.BYE, reason: body[0] };
        }
        default:
            throw new ControlError(`unknown opcode: 0x${op.toString(16)}`);
    }
}

/** Encode a HELLO_ACK control payload. */
export function encodeHelloAck({ protoVer = 1, accepted = 1, sessionId = 1 } = {}) {
    const out = new Uint8Array(7);
    out[0] = Opcode.HELLO_ACK;
    out[1] = protoVer;
    out[2] = accepted;
    out[3] = sessionId & 0xff;
    out[4] = (sessionId >>> 8) & 0xff;
    out[5] = (sessionId >>> 16) & 0xff;
    out[6] = (sessionId >>> 24) & 0xff;
    return out;
}

/** Encode an ADVERTISE_ACK. */
export function encodeAdvertiseAck(topicId, accepted = 1) {
    return Uint8Array.of(Opcode.ADVERTISE_ACK, topicId & 0xff, accepted & 0xff);
}

/** Encode a PONG with the given nonce. */
export function encodePong(nonce) {
    const out = new Uint8Array(5);
    out[0] = Opcode.PONG;
    out[1] = nonce & 0xff;
    out[2] = (nonce >>> 8) & 0xff;
    out[3] = (nonce >>> 16) & 0xff;
    out[4] = (nonce >>> 24) & 0xff;
    return out;
}

export const helpers = { bytesToHex };
